### PR TITLE
[runtime env] allow working_dir to be a zipped package

### LINKED ISSUE #20343: [runtime_env] Support a local .zip file for working_dir
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

{"working_dir": "my_prepackaged_files.zip"}

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by hckuo2**
## Why are these changes needed?
Check if working_dir is a zip, unzip it if so.

## Related issue number
Closes  #20343

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by hckuo** [comment]
This is what I came up. Could you provide some guidance for writing this test?

**Comment by hckuo** [comment]
Thinking about an alternative which expand `download_and_unpack_package` to handle local zip files. This may be a cleaner patch.

**Comment by architkulkarni** [comment]
I think for this to work, we need to modify `upload_working_dir_if_needed` (or maybe just modify `upload_package_if_needed`)?  The reason is that we need to upload the working_dir zip file from the local machine (the machine that the driver script is running on) to the head node.  

For local non-zipped directories, that's what currently happens in `upload_working_dir_if_needed`, and this function also replaces the `working_dir` field of the `runtime_env` with a URI to refer to the zipped working_dir package which is stored on the head node.  All the nodes in the cluster then download and unzip the working_dir from the head node using that URI.

For testing, perhaps you can take a look at the `tmp_working_dir` fixture and modify it or create a new one to use a local zip file instead?

**Comment by hckuo** [comment]
I unzip the package first to handle excludes and follow the original workflow. Would this work?

File: python/ray/_private/runtime_env/working_dir.py
Comment (architkulkarni) [inline python/ray/_private/runtime_env/working_dir.py:58]: If the kwarg `unlink_zip` is set to `True`, does that delete the user's `.zip` file?  That seems like unexpected behavior.

File: python/ray/_private/runtime_env/working_dir.py
Comment (architkulkarni) [inline python/ray/_private/runtime_env/working_dir.py:63]: After the package is uploaded, the unzipped working_dir will still be lying around on the user's local machine.  Should we delete it?

File: python/ray/_private/runtime_env/working_dir.py
Comment (architkulkarni) [inline python/ray/_private/runtime_env/working_dir.py:51]:
```suggestion
    except ValueError:  # working_dir is not a directory
```

File: python/ray/_private/runtime_env/working_dir.py
Comment (architkulkarni) [reply python/ray/_private/runtime_env/working_dir.py:63]: Also, should we do the unzipping in a temporary directory to avoid messing with the user's directory?  We should probably also delete the unzipped working_dir even in the case where `upload_package_if_needed` fails or throws an exception.  I think you can do both of these using https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory as a context manager.

**Comment by architkulkarni** [comment]
I think this should work, can you add or modify tests in `test_runtime_env_working_dir.py` or `test_runtime_env_working_dir_2.py` to test this case?

File: python/ray/_private/runtime_env/working_dir.py
Comment (hckuo) [reply python/ray/_private/runtime_env/working_dir.py:71]: Does it look good now?

**Comment by hckuo** [comment]
test added

**Comment by architkulkarni** [comment]
Looks good to me, can you please update the documentation as appropriate?  It's in `handling-dependencies.rst`.  Also I used the wrong label--let me try this one, maybe it'll allow you to remove it :)

**Comment by hckuo** [comment]
I still can not edit tags but I have updated the doc. Question: what's the difference between `py_modules` vs `working_dir`? It seems  to me that `working_dir` is sufficient to do what `py_modules` does.

**Comment by architkulkarni** [comment]
I think the advantage of `py_modules` over `working_dir` is that you can include a list of different modules that may lie in different directories.  Also, you can pass in the module object directly instead of just specifying the path to the package directory.

File: doc/source/handling-dependencies.rst
Comment (architkulkarni) [inline doc/source/handling-dependencies.rst:257]: Right now it looks like only (2) has the 100Mib limit, but can you clarify that it applies to both (1) and (2)?

**Comment by architkulkarni** [comment]
Looks like there are some merge conflicts unfortunately, could you please address those?

**Comment by hckuo** [comment]
Resolved conflicts :)

**Review by architkulkarni — APPROVED** [review]

**Comment by architkulkarni** [comment]
test_actor_client_mode  flaky on master
test_client_reconnect flaky on master
tests:test_multi_node_3 flaky on master
tests:test_runtime_env_complicated flaky on master

**Review by edoakes — DISMISSED** [review]
I don't think we should be doing this strange zip -> unzip -> zip again path. Can we instead just refactor the packaging code to take a zip file and calculate the hash + upload it directly?

I'm worried about maintainability here. I also think the performance hit and excludes behavior would be really surprising to users.

**Comment by architkulkarni** [comment]
@hckuo Sorry for the change here, I do agree that @edoakes's approach makes more sense and we should do that instead.

I think we can reuse the test you wrote.  The code changes might look something like this:  We can either refactor `upload_package_if_needed` to allow for files that are already zipped, or in `upload_working_dir_if_needed` we can generate the URI (by hashing the zip file) and then directly call `upload_package_to_gcs` like we currently do in `upload_package_if_needed`.  I think the second approach might be simpler, but maybe you can think of a better approach.

**Comment by hckuo** [comment]
In this case, the exclude flag would have no effect which I think could be surprising to users?

**Comment by architkulkarni** [comment]
Yeah, the excludes field would have no effect.  I'm not completely sure, but I think handling excludes for zip files is more surprising than not handling excludes.  I think of zip files as kind of a static artifact, so dynamically modifying its contents might be unexpected.  I agree it's not the best experience in the case where users have to repeatedly manually exclude files before zipping up their directory, but I think this use case is not as common.

**Comment by hckuo** [comment]
I see. That makes sense.

**Comment by hckuo** [comment]
@architkulkarni please review.

**Review by architkulkarni — APPROVED** [review]
Looks good!  @edoakes will review again after the holidays.

**Comment by architkulkarni** [comment]
Actually, looks like I can dismiss the change request.  @hckuo once you resolve the merge conflict, I think we can merge this!

File: python/ray/tests/test_runtime_env_working_dir.py
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_working_dir.py:46]: Very sorry for not noticing this earlier--could you please remove this?  `tmp_working_dir` is already defined in `tests/conftest.py`, and we shouldn't duplicate it.

Comment (hckuo) [reply python/ray/tests/test_runtime_env_working_dir.py:46]: No worries. Will do! Thanks

**Comment by architkulkarni** [comment]
@edoakes Could you please merge this if it looks good to you?

**Pull request merged**
